File: 4_trees_and_graph/4_4_check_balanced.py
import sys
sys.path.append("/Users/pauLi/Documents/Interviews/algo-practice/")

from bi_node import BiNode

def check_balanced(root):
  if root is None:
    return True
  else:
    return check_by_height(root) != -1

def check_by_height(root):
  if root.left_node is None and root.right_node is None:
    return 1
  left = 0 if root.left_node is None else check_by_height(root.left_node)
  right = 0 if root.right_node is None else check_by_height(root.right_node)
  if left == -1 or right == -1 or abs(left-right) > 1:
    return -1 
  else:
    return max(left, right) + 1

def test():
  t = None
  print('PASS' if check_balanced(t) else 'FAIL')

  t = BiNode(data=6)
  print('PASS' if check_balanced(t) else 'FAIL')

  t = BiNode(
    data=6,
    left_node=BiNode(data=7),
    right_node=BiNode(
      data=8,
      right_node=BiNode(data=9)
    )
  )
  print('PASS' if check_balanced(t) else 'FAIL')

  t = BiNode(
    data=6,
    left_node=BiNode(data=7),
    right_node=BiNode(
      data=9,
      left_node=BiNode(
        data=10,
        right_node=BiNode(data=11)
      ),
      right_node=BiNode(data=12)
    )
  )
  print('PASS' if not check_balanced(t) else 'FAIL')

  t = BiNode(
    data=6,
    left_node=BiNode(data=7),
    right_node=BiNode(
      data=9,
      left_node=BiNode(data=10),
      right_node=BiNode(
        data=12,
        left_node=BiNode(
          data=15,
          left_node=BiNode(data=11)
        )
      )
    )
  )
  print('PASS' if not check_balanced(t) else 'FAIL')

if __name__ == '__main__':
    test()

File: cracking/4_trees_and_graph/4.4_check_balanced.rb
load "#{Dir.home}/Documents/Interviews/algo-practice/bi_node.rb"

def check_balanced(root)
  return true if root.nil?
  check_by_height(root) != -1
end

def check_by_height(root)
  return 1 if root.left_node.nil? && root.right_node.nil?
  left = root.left_node.nil? ? 0 : check_by_height(root.left_node)
  right = root.right_node.nil? ? 0 : check_by_height(root.right_node)
  return -1 if left == -1 || right == -1 || (left-right).abs > 1
  return [left, right].max + 1
end

def test
  t = nil
  puts check_balanced(t) ? 'PASS' : 'FAIL'

  t = BiNode.new(data: 6)
  puts check_balanced(t) ? 'PASS' : 'FAIL'

  t = BiNode.new(
    data: 6,
    left_node: BiNode.new(data: 7),
    right_node: BiNode.new(
      data: 8,
      right_node: BiNode.new(data: 9)
    )
  )
  puts check_balanced(t) ? 'PASS' : 'FAIL'

  t = BiNode.new(
    data: 6,
    left_node: BiNode.new(data: 7),
    right_node: BiNode.new(
      data: 9,
      left_node: BiNode.new(
        data: 10,
        right_node: BiNode.new(data: 11)
      ),
      right_node: BiNode.new(data: 12)
    )
  )
  puts !check_balanced(t) ? 'PASS' : 'FAIL'

  t = BiNode.new(
    data: 6,
    left_node: BiNode.new(data: 7),
    right_node: BiNode.new(
      data: 9,
      left_node: BiNode.new(data: 10),
      right_node: BiNode.new(
        data: 12,
        left_node: BiNode.new(
          data: 15,
          left_node: BiNode.new(data: 11)
        )
      )
    )
  )
  puts !check_balanced(t) ? 'PASS' : 'FAIL'
end

